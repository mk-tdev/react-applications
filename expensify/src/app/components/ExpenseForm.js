import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocued: false,
      errorText: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;

    this.setState(() => ({
      description,
    }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  };

  onAmountChange = ({ target }) => {
    const amount = target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChanged = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocuesChanged = ({ focused }) => {
    this.setState(() => ({
      calendarFocued: focused,
    }));
  };

  onFormSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description && !this.state.amount) {
      this.setState(() => ({ errorText: "Please provide proper data!" }));
    } else {
      this.setState(() => ({ errorText: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.errorText && (
          <p className="errorText">{this.state.errorText}</p>
        )}
        <form onSubmit={this.onFormSubmit}>
          <div>
            <input
              type="text"
              autoFocus
              placeholder="Description"
              value={this.state.description}
              onChange={this.onDescriptionChange}
            />
          </div>
          <div>
            <input
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            />
          </div>
          <div>
            <textarea
              placeholder="Add a note about it"
              value={this.state.note}
              onChange={this.onNoteChange}
            ></textarea>
          </div>

          <div>
            <SingleDatePicker
              date={this.state.createdAt}
              onDateChange={this.onDateChanged}
              focused={this.state.calendarFocued}
              onFocusChange={this.onFocuesChanged}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>

          <div>
            <button>
              {this.props.expense ? "Edit Expense" : "Add Expense"}
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
