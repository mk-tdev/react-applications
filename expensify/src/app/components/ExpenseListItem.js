import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";

import { startRemoveExpenses } from "../actions/expenses";

export const ExpenseListItem = ({
  description,
  amount,
  createdAt,
  id,
  dispatch,
}) => {
  // console.log("moment", moment(createdAt).date());

  return (
    <div className="expense-item">
      <div>
        <Link to={`/edit/${id}`}>
          <h4>{description}</h4>
        </Link>
        <p>
          ${amount} - (at: {createdAt})
        </p>
      </div>
      <button
        onClick={() => {
          dispatch(startRemoveExpenses({ id }));
        }}
      >
        Remove
      </button>
    </div>
  );
};

export default connect()(ExpenseListItem);
