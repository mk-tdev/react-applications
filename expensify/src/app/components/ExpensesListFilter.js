import React from "react";
import { connect } from "react-redux";

import { setTextFilter, sortByAmount, sortByDate } from "../actions/filters";

export class ExpensesListFilter extends React.Component {
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpensesListFilter);
