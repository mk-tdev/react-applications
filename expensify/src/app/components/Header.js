import React, { useEffect } from "react";
import { Link, NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout, auth, history }) => {
  useEffect(() => {
    if (auth) {
      if (history.location.pathname === "/") history.push("/home");
    } else {
      history.push("/");
    }
  }, [auth]);

  return (
    <header className="header">
      <span>Expensify</span>

      {auth ? (
        <div>
          <NavLink to="/home" activeClassName="is-active" exact>
            Home
          </NavLink>
          <NavLink to="/create" activeClassName="is-active">
            Create
          </NavLink>
          <button onClick={startLogout}>Logout</button>
        </div>
      ) : null}

      {/* <NavLink to="/edit" activeClassName="is-active">
        Edit
      </NavLink> */}
      {/* {auth && <button onClick={startLogout}>Logout</button>} */}
      {/* <NavLink to="/help" activeClassName="is-active">
        Help
      </NavLink> */}
    </header>
  );
};

const mapStateToProps = (state, props) => {
  return {
    auth: state.auth.isAuthenticated,
  };
};

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Header));
