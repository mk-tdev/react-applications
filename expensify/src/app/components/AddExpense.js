import React from "react";
import { connect } from "react-redux";
// import { useHistory } from "react-router-dom";
import { withRouter } from "react-router-dom";

import ExpenseForm from "./ExpenseForm";
import { initAddExpense } from "../actions/expenses";

export class AddExpense extends React.Component {
  onSubmit = (expense) => {
    this.props.initAddExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToPros = (dispatch) => ({
  initAddExpense: (expense) => dispatch(initAddExpense(expense)),
});

export default withRouter(connect(undefined, mapDispatchToPros)(AddExpense));
