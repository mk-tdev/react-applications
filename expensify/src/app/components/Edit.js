import React from "react";
import { connect } from "react-redux";
// import { useHistory } from "react-router-dom";
import { withRouter } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class Edit extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h2>Edit</h2>

        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />

        <div>
          <button onClick={this.onRemove}>Remove This Expense</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((exp) => exp.id === props.match.params.id),
  };
};

const mapDispatchToPros = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense(id)),
});

export default withRouter(connect(mapStateToProps, mapDispatchToPros)(Edit));
