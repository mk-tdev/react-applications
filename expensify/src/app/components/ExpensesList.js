import React from "react";

import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

function ExpensesList(props) {
  return (
    <div>
      {props.expenses.map((exp) => {
        return <ExpenseListItem key={exp.id} {...exp} />;
      })}
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpensesList);
