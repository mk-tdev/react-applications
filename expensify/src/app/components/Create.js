import React from "react";
import AddExpense from "./AddExpense";

function Create() {
  return <AddExpense />;
}

export default Create;
