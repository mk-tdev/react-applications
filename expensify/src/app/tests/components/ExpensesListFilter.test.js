import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
configure({ adapter: new Adapter() });

import { ExpensesListFilter } from "../../components/ExpensesListFilter";
import { filters, altFilters } from "../fixtures/filters";

let setTextFilter, sortByAmount, sortByDate, wrapper;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByAmount = jest.fn();
  sortByDate = jest.fn();
  wrapper = shallow(
    <ExpensesListFilter
      filters={filters}
      sortByDate={sortByDate}
      sortByAmount={sortByAmount}
      setTextFilter={setTextFilter}
    />
  );
});

test("should render ExpensesListFilter", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesListFilter with different filters", () => {
  wrapper.setProps({
    filters: altFilters,
  });
  expect(wrapper).toMatchSnapshot();
});

test("should update state when text changes", () => {
  wrapper
    .find("input")
    .at(0)
    .simulate("change", { target: { value: "PUB" } });
  expect(setTextFilter).toHaveBeenLastCalledWith("PUB");
});

test("should update state when select changes", () => {
  wrapper
    .find("select")
    .at(0)
    .simulate("change", { target: { value: "date" } });
  expect(sortByDate).toHaveBeenCalled();
});

test("should update state when select changes", () => {
  wrapper
    .find("select")
    .at(0)
    .simulate("change", { target: { value: "amount" } });
  expect(sortByAmount).toHaveBeenCalled();
});
