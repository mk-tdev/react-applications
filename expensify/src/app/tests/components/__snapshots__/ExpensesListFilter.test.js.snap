// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesListFilter 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
</div>
`;

exports[`should render ExpensesListFilter with different filters 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value="rent"
  />
  <select
    onChange={[Function]}
    value="amout"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
</div>
`;
